### PR TITLE
Capture User and Case Manager Contact Details (take 2)

### PR DESCRIPTION
**Pull request by davidatkinsuk**
Relates to https://dsdmoj.atlassian.net/browse/APS-410

Note - this was previously put out for review on https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1566. To make rewriting the history easier to manage, i've resolved these issues on a new branch

Previously, the overridden user contact details and case manager contact details were only captured in the application data JSON. This PR introduces a cas_1_application_user_details table to contain this information, and links to it from the approved_premises_applications table. We also introduce a boolean case_manager_is_not_applicant to capture if the case manager is not the applicant.

A migration (backfill) job is also added to populate this data in prod/pre-prod from the data JSO

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt [hunk 1]
```diff
@@ -1646,6 +1690,178 @@ class ApplicationServiceTest {
...
+        applicationId,
+        submitApprovedPremisesApplication,
+        username,
+        "jwt",
+        user.probationRegion.id,
+      )
+
+      assertThat(result is AuthorisableActionResult.Success).isTrue
+
+      verify { mockCas1ApplicationUserDetailsRepository.delete(existingCaseManagerUserDetails) }
```
Comment (edavey) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt:1833]: Do we also want an integration test to verify that the API now exposes `caseManagerUserDetails:  null` when the `caseManagerIsNotApplicant` flag is toggled to false? Or is it sufficient to know that the service asks the repository to delete the `Cas1ApplicationUserDetailsEntity` record?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/migration/Cas1UserDetailsMigrationJob.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    val applicantNameOverridden = detailsUptoDateMap?.contains("name") ?: false
+    val applicantEmailOverridden = detailsUptoDateMap?.contains("emailAddress") ?: false
+    val applicantPhoneOverridden = detailsUptoDateMap?.contains("phoneNumber") ?: false
+
+    val applicantName = if (applicantNameOverridden) {
+      if (confirmYourDetailsMap!!.contains("name")) {
+        confirmYourDetailsMap["name"]!! as String
+      } else {
+        error("applicant name marked as overridden, but no value defined for 'confirm-your-details/name' for application ${application.id}")
```
Comment (edavey) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/migration/Cas1UserDetailsMigrationJob.kt:122]: this looks robust, it will be interesting to see if you've foreseen all the possible issues!

**Review by edavey — APPROVED** [review]
This looks great.

I may have missed it but I'm not sure if we need an integration test for the scenario where case manager details are recorded and then need to be nullified?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt [hunk 3]
```diff
@@ -1538,6 +1539,56 @@ class ApplicationServiceTest {
...
+        "jwt",
+        null,
+      )
+
+      assertThat(result is AuthorisableActionResult.Success).isTrue
+      result as AuthorisableActionResult.Success
+      assertThat(result.entity is ValidatableActionResult.GeneralValidationError).isTrue
+      val validatableActionResult = result.entity as ValidatableActionResult.GeneralValidationError
+
+      assertThat(validatableActionResult.message).isEqualTo("caseManagerIsNotApplicant must be provided if caseManagerIsNotApplicant is true")
```
[the remark this answers is not in the document]
Comment (davidatkinsuk) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt]: oops, fixed

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt (hunk 1 above)
Comment (davidatkinsuk) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt:1833]: I think unit test is sufficient in this case
